### PR TITLE
Upgrade to rails `6.0.4.6`

### PR DESCRIPTION
**Pull request by waldofouche**
Version chosen based on: https://endoflife.date/rails

I believe this upgrade works? All the tests passed locally and the application boots:

```
9 deprecation warnings total

Finished in 5 minutes 31 seconds (files took 5.23 seconds to load)
2085 examples, 0 failures

➜  saml-service git:(upgrades/rails) ✗ rails s
=> Booting Puma
=> Rails 6.1.4.4 application starting in development
=> Run `bin/rails server --help` for more startup options
[51372] Puma starting in cluster mode...
[51372] * Puma version: 5.6.2 (ruby 2.7.4-p191) ("Birdie's Version")
[51372] *  Min threads: 1
[51372] *  Max threads: 1
[51372] *  Environment: development
[51372] *   Master PID: 51372
[51372] *      Workers: 3
[51372] *     Restarts: (✔) hot (✖) phased
[51372] * Preloading application
[51372] * Listening on http://0.0.0.0:8080
[51372] Use Ctrl-C to stop
[51372] - Worker 0 (PID: 51411) booted in 0.0s, phase: 0
[51372] - Worker 1 (PID: 51412) booted in 0.0s, phase: 0
[51372] - Worker 2 (PID: 51413) booted in 0.0s, phase: 0
Started GET "//api/discovery/entities" for 127.0.0.1 at 2022-02-23 11:46:25 +0800
```

Also checked `http://localhost:8080//api/discovery/entities`

Screen Shot 2022-02-23 at 11 35 08 am

Part of https://github.com/ausaccessfed/stockpile/issues/101

### CODE REVIEW

**Review by ArthurZheng — APPROVED** [review]
👍 nice work

**Review by matthew-puku — APPROVED** [review]
Sweet 👍

**Comment by rianniello** [comment]
https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-5-2-to-rails-6-0
https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-0-to-rails-6-1
🤔🤔🤔🤔

**Comment by rianniello** [comment]
Perhaps worth having a quick read through those ^

**Comment by waldofouche** [comment]
My attempts at enabling `zeitwerk` loader per https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading and https://www.fastruby.io/blog/rails/upgrade/zeitwerk/upgrading-to-zeitwerk.html are not going too well, Only causing walls of errors: 

```
SEQUEL DEPRECATION WARNING: Using a joined dataset as a Sequel::Model dataset is deprecated and will be removed in Sequel 5.  Use the class_table_inheritance plugin :alias option in LocalizedURI.
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/sequel-4.49.0/lib/sequel/model/base.rb:1027:in `convert_input_dataset'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/sequel-4.49.0/lib/sequel/model/base.rb:885:in `set_dataset'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/sequel-4.49.0/lib/sequel/plugins/class_table_inheritance.rb:318:in `block in inherited'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/sequel-4.49.0/lib/sequel/plugins/class_table_inheritance.rb:301:in `instance_eval'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/sequel-4.49.0/lib/sequel/plugins/class_table_inheritance.rb:301:in `inherited'
/Users/waldofouche/dev/saml-service/app/models/mdui/information_url.rb:4:in `<module:MDUI>'
/Users/waldofouche/dev/saml-service/app/models/mdui/information_url.rb:3:in `<top (required)>'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:27:in `require'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:27:in `require'
/Users/waldofouche/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:95:in `const_get'
```

The closest I got to zeitwerk loading working was:

```
➜  saml-service git:(upgrades/rails) ✗ bin/rails zeitwerk:check
Hold on, I am eager loading the application.

SEQUEL DEPRECATION WARNING: Using class_table_inheritance with duplicate column names (RequestedAttribute => [:created_at, :updated_at]) in subclass tables (other than the primary key column)', 'Make sure all tables used have unique column names, or implement support for handling duplicate column names in the class_table_inheritance plugin

<INSERT ABOUT 400 more lines of errors related to the sequel gem>

WARNING: The following directories will only be checked if you configure
them to be eager loaded:

  /Users/waldofouche/dev/saml-service/app/jobs/concerns

You may verify them manually, or add them to config.eager_load_paths
in config/application.rb and run zeitwerk:check again.

Otherwise, all is good!
```

And ofcourse all the tests fail

**Comment by waldofouche** [comment]
also facing issues similar to  https://github.com/rails/rails/issues/40904

**Comment by waldofouche** [comment]
Hmm, this is gigantic, it is probabaly easiest to review the commits in order, than look at the crazy diff

**Comment by waldofouche** [comment]
Initially intended to upgrade to `6.1.4.4` in this PR, however there were too many changes for one PR, have split the branches and created a seperate PR https://github.com/ausaccessfed/saml-service/pull/267 for easier review

File: app/jobs/update_entity_source.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@
 require 'metadata/schema'
 require 'net/http'
 
-class UpdateEntitySource
-  include SetSAMLTypeFromXML
+class UpdateEntitySource <
```
Comment (matthew-puku) [inline app/jobs/update_entity_source.rb:6]: This doesn't look right 🤔. Was it intentional?

```suggestion
class UpdateEntitySource
```

**Review by matthew-puku — APPROVED** [review]

File: app/jobs/update_entity_source.rb (hunk 1 above)
Comment (waldofouche) [reply app/jobs/update_entity_source.rb:6]: Ah that was me playing around with the `includes` and stuff, was not intended to commit those changes.

**Comment by rianniello** [comment]
@vladimir-mencl-eresearch would like your input on this one please.

**Review by rianniello — APPROVED** [review]
LGTM, Great work, appreciate the detail.

Did some rudimentary tests to help smoothen the rollout:
```xml
$ http http://localhost:8080/mdq/JMWhTb1X01EB_4gZ0zUfOQ/entities Accept:application/samlmetadata+xml
HTTP/1.1 200 OK
Cache-Control: max-age=21600, private
Content-Type: application/samlmetadata+xml; charset=utf-8
Date: Mon, 07 Mar 2022 04:53:45 GMT
ETag: 3fc3808f9d3085e7aeae62ff2ab92312
Last-Modified: Mon, 07 Mar 2022 04:51:41 -0000
Referrer-Policy: strict-origin-when-cross-origin
Transfer-Encoding: chunked
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-Request-Id: 8329f444-ba73-4b7a-b8d2-89ccd25d9e8d
X-Runtime: 0.009469
X-XSS-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
<EntitiesDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:idpdisc="urn:oasis:names:tc:SAML:profiles:SSO:idp-discovery-protocol" xmlns:mdrpi="urn:oasis:names:tc:SAML:metadata:rpi" xmlns:mdui="urn:oasis:names:tc:SAML:metadata:ui" xmlns:mdattr="urn:oasis:names:tc:SAML:metadata:attribute" xmlns:shibmd="urn:mace:shibboleth:metadata:1.0" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:fed="http://docs.oasis-open.org/wsfed/federation/200706" xmlns:privacy="http://docs.oasis-open.org/wsfed/privacy/200706" xmlns:remd="http://refeds.org/metadata" ID="dicta20220307045329" Name="gerhold.biz" validUntil="2022-03-07T05:53:29Z">
  <ds:Signature>
    <ds:SignedInfo>
      <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
      <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
      <ds:Reference URI="#dicta20220307045329">
        <ds:Transforms>
          <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
          <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
        </ds:Transforms>
        <ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
        <ds:DigestValue>kJ0Opk+1ZrMH2rSJa03HeTmqOY2qO5Leh58CjwKd7pk=</ds:DigestValue>
      </ds:Reference>
    </ds:SignedInfo>
    <ds:SignatureValue>lqebcYlpwuAr+ZoFkJiwy25MP0zb9z7rdnZlWLVgJ3sUmx3ooow8filMN/jix5YXod9WwH2LtDcoA8AWhmYniVIK2pZPHAWput40ZwWlvF9KMH0dekw4t6oydLsu0AJjgXhUm/JK29+ccqhTtczkqqWu/hhzaThTKKwsxUv8jpbdHgLX+E5G6EgMiNaABWdi51PxD39i1RuPywROFXl4lsTsi7g1apqA6J9TsJrFWAJm2bMfrvERNwYOZSVrzfR6qzNmdBRMlGAXsDQcU9Ag78DI4+p7b57FgWuoxJfgGk0Fwzh1akYLO07XiXhkYDNl+CwPlrB1MlSav5V3yGW7vg==</ds:SignatureValue>
    <ds:KeyInfo>
      <ds:KeyValue>
        <ds:RSAKeyValue>
          <ds:Modulus>yXYjRR91RFNranjk6FrIdHuSLFeGFGH7iJ8C0nvI8ki9Qr11HO7KZSAU+VLGGVgP4VgOBCm+SHK+7Lfx0atA7UBh4yQClCtBsZjyefNdu9vN66X+XhhuU54pTuwrWnKSVDR3z9LHeNkRg9BlhKnvDi0SBS7/Feoa495f5yqNKLg5MtkVDk3oFxP3UDQzt0yaLypYUkNcEDXLOBnR/+E9NSgsvnZLWTkqaGFI+EHjw4CkrUKHQmEpz7EAtMQW28+VUdNj7W7DxtU+qm6fpi6ABDS8bBWQuNxCz5seP8YEJlxRGosgzGmtb7C3ifhf+4iZPMHHaOyymISta2WNp1lsrQ==</ds:Modulus>
          <ds:Exponent>AQAB</ds:Exponent>
        </ds:RSAKeyValue>
      </ds:KeyValue>
      <ds:X509Data>
        <ds:X509Certificate>MIICuzCCAaOgAwIBAgIBADANBgkqhkiG9w0BAQsFADAhMR8wHQYDVQQDDBZYRUlt
dVhqSUd6OENSaUFnREEtWHZBMB4XDTIyMDMwNzA0NDE1MloXDTIyMDMwNzA1NDE1
MlowITEfMB0GA1UEAwwWWEVJbXVYaklHejhDUmlBZ0RBLVh2QTCCASIwDQYJKoZI
hvcNAQEBBQADggEPADCCAQoCggEBAMl2I0UfdURTa2p45OhayHR7kixXhhRh+4if
AtJ7yPJIvUK9dRzuymUgFPlSxhlYD+FYDgQpvkhyvuy38dGrQO1AYeMkApQrQbGY
8nnzXbvbzeul/l4YblOeKU7sK1pyklQ0d8/Sx3jZEYPQZYSp7w4tEgUu/xXqGuPe
X+cqjSi4OTLZFQ5N6BcT91A0M7dMmi8qWFJDXBA1yzgZ0f/hPTUoLL52S1k5Kmhh
SPhB48OApK1Ch0JhKc+xALTEFtvPlVHTY+1uw8bVPqpun6YugAQ0vGwVkLjcQs+b
Hj/GBCZcURqLIMxprW+wt4n4X/uImTzBx2jsspiErWtljadZbK0CAwEAATANBgkq
hkiG9w0BAQsFAAOCAQEAqtVgRlhOdJXQQN7qUaIU7djGhbQ8CD6DtbLOfz5H2YIu
EAaNpGJORRR+7kcjJ88uAFAMqPvJpUiNY7f6LhHEbhWladaglHql3rtNRqKnjE2c
Sb+sdtR4O4Qdn53Kzrp2FFFqTVPhGMo1kDBkYQvv5qwbzNcVuD8/Bp5L10uzQ2Id
cA+ehiHQRdvMvb/4ihiOCuevJ52E0lA/2P89+vKLb3zb833hh69ljbdxUhjomJY3
ZLZkZmcvLDwNRiFs7sVEliu0nuBB6Yi/m6gLDGMU8mjNa0GHn515xL1Cc83g5suU
/3iAtwDcPYyHKpRyZXAeiZYMrpaDzF5DQKvtRYZlgg==</ds:X509Certificate>
      </ds:X509Data>
    </ds:KeyInfo>
  </ds:Signature>
  <Extensions>
    <mdrpi:PublicationInfo publisher="http://lehner-greenholt.co/galen" creationInstant="2022-03-07T04:53:29Z" publicationId="dicta20220307045329">
      <mdrpi:UsagePolicy xml:lang="en">http://www.edugain.org/policy/metadata-tou_1_0.txt</mdrpi:UsagePolicy>
    </mdrpi:PublicationInfo>
  </Extensions>
</EntitiesDescriptor>
```

```
$ curl http://localhost:8080/api/discovery/entities
{"identity_providers":[],"service_providers":[]}%
```

**Comment by rianniello** [comment]
Also little more testing
```ruby
[4] pry(main)> FactoryBot::create :idp_sso_descriptor
=> #<IDPSSODescriptor @values={:id=>1, :entity_descriptor_id=>1, :organization_id=>nil, :error_url=>nil, :extensions=>nil, :created_at=>2022-03-07 04:57:40 UTC, :updated_at=>2022-03-07 04:57:40 UTC, :kind=>"IDPSSODescriptor", :enabled=>true, :want_authn_requests_signed=>false}>
[5] pry(main)> FactoryBot::create :sp_sso_descriptor
=> #<SPSSODescriptor @values={:id=>2, :entity_descriptor_id=>2, :organization_id=>nil, :error_url=>nil, :extensions=>nil, :created_at=>2022-03-07 04:57:44 UTC, :updated_at=>2022-03-07 04:57:44 UTC, :kind=>"SPSSODescriptor", :enabled=>true, :authn_requests_signed=>false, :want_assertions_signed=>false}>
[6] pry(main)> FactoryBot::create :sp_sso_descriptor
=> #<SPSSODescriptor @values={:id=>3, :entity_descriptor_id=>3, :organization_id=>nil, :error_url=>nil, :extensions=>nil, :created_at=>2022-03-07 04:57:45 UTC, :updated_at=>2022-03-07 04:57:45 UTC, :kind=>"SPSSODescriptor", :enabled=>true, :authn_requests_signed=>false, :want_assertions_signed=>false}>
```

```
$ curl http://localhost:8080/api/discovery/entities
```
```json
{"identity_providers":[{"entity_id":"http://schmidt-roberts.biz/moses/shibboleth","tags":[],"names":[],"logos":[],"descriptions":[],"single_sign_on_endpoints":{"soap":[]},"geolocations":[],"domains":[]}],"service_providers":[{"entity_id":"http://rutherford.net/reynalda/shibboleth","tags":[],"names":[],"logos":[],"descriptions":[],"all_discovery_response_endpoints":[],"information_urls":[],"privacy_statement_urls":[]},{"entity_id":"http://wisoky.co/sheldon/shibboleth","tags":[],"names":[],"logos":[],"descriptions":[],"all_discovery_response_endpoints":[],"information_urls":[],"privacy_statement_urls":[]}]}%
```

**Comment by vladimir-mencl-eresearch** [comment]
> @vladimir-mencl-eresearch would like your input on this one please.

Sure, happy to.

Sorry, this slipped in backlog of GitHub notifications and only found it now - will look at the PR tomorrow.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Hi @rianniello ,

Thanks for the opportunity to comment.

I've just gone through the commits and it all makes sense to me.

I've also tried rolling this out in our DEV environment and all works as expected (after merging with our changes).

My only concern (already expressed in ausaccessfed/reporting-service#253) is the way the Rails version is expressed in `Gemfile` as `gem 'rails', '6.0.4.4'`.  I'd expect pinning to exact version happens only in `Gemfile.lock` and here it would be expressed as:
```
gem 'rails', '<6.1'
```

Otherwise looks all good to me.

Cheers,
Vlad

**Comment by waldofouche** [comment]
Thanks for the feedback @vladimir-mencl-eresearch, I have fixed up the version pinning as recommended

**Comment by vladimir-mencl-eresearch** [comment]
> Thanks for the feedback @vladimir-mencl-eresearch, I have fixed up the version pinning as recommended

.... and I see the `bundle update` is in as well - thanks!

**Comment by rianniello** [comment]
Brilliant, appreciate testing the rollout @vladimir-mencl-eresearch. We hope to be able to do this easily soon too. This should be good to merge in the AM next week @waldofouche

**Comment by waldofouche** [comment]
> This should be good to merge in the AM next week @waldofouche

Sounds good! I will merge early monday morning and keep an eye on the rollout

**Pull request merged**